### PR TITLE
[WIP] update pl=0.8.5

### LINKED ISSUE #5739: TypeError: join() argument must be str or bytes, not 'NoneType'
Trying to run seq2seq example scripts with multiple gpus and wandb logging and I get
```
Traceback (most recent call last):
  File "/home/martongyorgy/projects/tmp/transformers/examples/seq2seq/finetune.py", line 364, in <module>
	xsum_rouge.json
  File "/home/martongyorgy/projects/tmp/transformers/examples/seq2seq/finetune.py", line 342, in main
    logger=logger,
  File "/home/martongyorgy/projects/tmp/transformers/examples/lightning_base.py", line 330, in generic_train
    trainer.fit(model)
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 891, in fit
    self.ddp_train(task, model)
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/site-packages/pytorch_lightning/trainer/distrib_data_parallel.py", line 530, in ddp_train
    self.run_pretrain_routine(model)
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1046, in run_pretrain_routine
    self.configure_checkpoint_callback()
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/site-packages/pytorch_lightning/trainer/callback_config.py", line 60, in configure_checkpoint_callback
    "checkpoints"
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/home/martongyorgy/miniconda3/envs/transformers/lib/python3.7/genericpath.py", line 153, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'NoneType'
```

To reproduce follow the instructions in the seq2seq example Readme to download XSUM then run
```
export PYTHONPATH="../":"${PYTHONPATH}"

python finetune.py \
    --learning_rate=3e-5 \
    --fp16 \
    --gpus 2 \
    --do_train \
    --n_val 1000 \
    --val_check_interval 0.1 \
    --data_dir xsum \
    --output_dir xsum_frozen_embs \
    --model_name_or_path t5-small \
    --train_batch_size 16 --eval_batch_size 16 --freeze_embeds --freeze_encoder \
    --num_train_epochs 6 \
    --max_target_length=60 --val_max_target_length=60 --test_max_target_length=100 \
    --logger wandb
```

Happens in pytorch-lightning 0.8.1, fixed in 0.8.4. But there is another problem in  0.8.4, see #5584

### PR DESCRIPTION
**Pull request by williamFalcon**
1. update to use from_argparse_args

2. deleted args that were manual but trainer has already.

3. fixed optimizer_step.

4. moved the LR logging to the correct hook

5. dropped the rank_zero_only stuff since that's not needed. All loggers and print ops happen only on rank_zero already. This was redundant.

6. proper use of setup()

@sshleifer let me know how I can test these changes

### CODE REVIEW

**Comment by williamFalcon** [comment]
@sshleifer this looks good now :) 

```
Epoch 1:   0%|                                                                                                                         | 2/13382 [00:01<3:15:14,  1.14it/s, loss=171200.000, v_num=1t6n9wkw]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8192.0
Epoch 1:   0%|                                                                                                                         | 3/13382 [00:02<2:28:51,  1.50it/s, loss=135029.328, v_num=1t6n9wkw]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Epoch 1:   0%|                                                                                                                         | 4/13382 [00:02<2:06:06,  1.77it/s, loss=109544.000, v_num=1t6n9wkw]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 2048.0
Epoch 1:  20%|███████████████████████▉                                                                                                | 2672/13382 [23:08<1:32:43,  1.93it/s, loss=7298.400, v_num=1t6n9wkw[[AEpoch 1:  50%|████████████████████████████████████████████████████████████▎                                                            | 6674/13382 [55:41<55:58,  2.00it/s, loss=28663.199, v_num=1t6n9wkw]
Epoch 1:  50%|████████████████████████████████████████████████████████████▎                                                            | 6675/13382 [55:46<56:02,  1.99it/s, loss=28663.199, v_num=1t6n9wkw]
Validating:  76%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                     | 48/63 [03:33<01:07,  4.52s/it]
Epoch 1:  53%|████████████████████████████████████████████████████████████████▎                                                        | 7114/13382 [59:07<52:05,  2.01it/s, loss=28257.600, v_num=1t6n9wkw]
```


**This PR:**
- Fixes the lightning_base.py
- adds best practices to it
- fixes seq2seq example

**TODO:**
- fix the other examples (separate PR likely?)

**Comments**
- There is a lot of flag mapping that kind of starts to limit what users can do. (accumulate_grad_batches vs whatever). It's not the worst thing, but it adds unnecessary points of failure.

- data: Download data in prepare_data, use setup to split/etc and assign.

**Comment by sshleifer** [comment]
- Do you have a ROUGE-2 score for that run?
- Does multi-gpu + `trainer.test` work?

**Comment by sshleifer** [comment]
I think you also need to fix examples/seq2seq/test_seq2seq_examples.py
and `examples/requirements.txt`. But I would only do that after you've verified
that val-rouge2 doesn't get worse on single and multi-gpu. You can see metrics in `output_dir/metrics.json`.

**Review by sshleifer — CHANGES_REQUESTED** [review]
Left comments w.r.t metrics inline, the code looks fine.

**Comment by williamFalcon** [comment]
@sshleifer ok, this is verified to work with 2 GPUs and apex... what do you want to do now?

File: examples/lightning_base.py [hunk 1]
```diff
@@ -251,26 +244,23 @@ def add_generic_args(parser, root_dir) -> None:
     parser.add_argument(
         "--fp16_opt_level",
         type=str,
-        default="O1",
+        default="O2",
```
Comment (sshleifer) [inline examples/lightning_base.py:247]: Is there a good reason to change this?

File: examples/lightning_base.py [hunk 2]
```diff
@@ -296,38 +289,28 @@ def generic_train(
 
     train_params = {}
 
+    # TODO: remove with PyTorch 1.6 since pl uses native amp
     if args.fp16:
-        train_params["use_amp"] = args.fp16
+        train_params["precision"] = 16
         train_params["amp_level"] = args.fp16_opt_level
 
-    if args.n_tpu_cores > 0:
```
Comment (sshleifer) [inline examples/lightning_base.py:303]: does TPU also work now?

**Review by sshleifer — COMMENTED** [review]
Cool. 
I think we need to update some bash scripts and unit tests to reflect new command line args, like `-logger_name`.
Otherwise LGTM. Also would be nice to get a rouge score from a full epoch.

**Comment by sshleifer** [comment]
You also need to run `make style` to satisfy the check_code_quality CI job.

**Comment by williamFalcon** [comment]
@sshleifer 0.8.5 is live. we can merge and finish this now :) 
Then update the rest of the examples and we should be golden.

**Comment by sshleifer** [comment]
A very similar change was merged @moscow25

**Pull request closed without merging**